revolv/payments/signals.py

```diff
@@ -0,0 +1,146 @@
+    funds for this AdminReinvestment.
+    """
+    init_kwargs = kwargs.get('kwargs')
+    if not init_kwargs:
+        raise NotEnoughFundingException()
+    invest_amount = init_kwargs.get('amount') or 0.0
+
+    global_repay_amount = AdminRepayment.objects.aggregate(
+        Sum('amount')
+    )['amount__sum'] or 0.0
+    global_reinvest_amount = AdminReinvestment.objects.aggregate(
+        Sum('amount')
+    )['amount__sum'] or 0.0
+    global_reinvest_pool = global_repay_amount - global_reinvest_amount
+    if global_reinvest_pool - invest_amount < 0.0:
+        raise NotEnoughFundingException()
+
+
+@receiver(signals.post_save, sender=AdminReinvestment)
+def post_save_admin_reinvestment(**kwargs):
+    """
+    When an AdminReinvestment is saved, we pool as many donors as we need to
+    fund the reinvestment, prioritizing users that have a preference for the
+    Category of the project begin invested into. We only consider users that
+    have a non-zero pool of investable money.
+
+    !!! TODO: actually prioritize by Category
+    """
+    instance = kwargs.get('instance')
+    total_left = instance.amount
+    pending_reinvestors = []
+    for user in RevolvUserProfile.objects.filter(reinvest_pool__gt=0.0):
+        total_left -= user.reinvest_pool
+        reinvest_amount = user.reinvest_pool + min(0.0, total_left)
+        pending_reinvestors.append((user, reinvest_amount))
+        if total_left <= 0.0:
+            break
+    for (user, amount) in pending_reinvestors:
+        reinvestment = Payment(user=user,
+                               project=instance.project,
+                               entrant=instance.admin,
+                               payment_type=PaymentType.objects.get_reinvestment(),
+                               admin_reinvestment=instance,
+                               amount=amount
+                               )
+        reinvestment.save()
+
+
+@receiver(signals.post_save, sender=Repayment)
+def post_save_repayment(**kwargs):
+    """
+    When a Repayment is saved, we increment the reinvest_pool in the related
+    user.
+    """
+    instance = kwargs.get('instance')
+    instance.user.reinvest_pool += instance.amount
+    instance.user.save()
+
+
+@receiver(signals.pre_delete, sender=Repayment)
+def pre_delete_repayment(**kwargs):
+    """
+    Before a Repayment is deleted, we decrement the reinvest_pool in the related
+    user.
+    """
+    instance = kwargs.get('instance')
+    instance.user.reinvest_pool -= instance.amount
+    instance.user.save()
+
+
+@receiver(signals.post_save, sender=Payment)
+def post_save_payment(**kwargs):
+    """
+    If the payment is organic, we add this payment's user as a donor to the
+    related project. If the payment is a reinvestment, we decrement the
+    reinvest_pool in the related user.
+    """
+    instance = kwargs.get('instance')
+    if instance.payment_type == PaymentType.objects.get_paypal():
```

Comment:
Supposedly we should compare whether the user == the entrant, right? Because when we put in paying with paypal directly instead of paying with CC processed via paypal, we'll have to change this (because both types will be considered an organic donation).

Even better would be to write a `Payment.isOrganic` method haha